internal/servers/controller/session_cleanup_job.go

```diff
@@ -0,0 +1,191 @@
+		statusGracePeriod: statusGracePeriod,
+	}, nil
+}
+
+// Name returns a short, unique name for the job.
+func (j *sessionCleanupJob) Name() string { return "sessionCleanupJob" }
+
+// Description returns the description for the job.
+func (j *sessionCleanupJob) Description() string {
+	return "Clean up session connections from disconnected workers"
+}
+
+// NextRunIn returns the next run time after a job is completed.
+//
+// The next run time is defined for sessionCleanupJob as one second.
+// This is because the job should run continuously to terminate
+// connections as soon as a worker has not reported in for a long
+// enough time. Only one job will ever run at once, so there is no
+// reason why it cannot run again immediately.
+func (j *sessionCleanupJob) NextRunIn() (time.Duration, error) { return time.Second, nil }
+
+// Status returns the status of the running job.
+func (j *sessionCleanupJob) Status() scheduler.JobStatus {
+	return scheduler.JobStatus{
+		Completed: j.numProcessed,
+		Total:     j.numWorkers,
+	}
+}
+
+// Run executes the job.
+func (j *sessionCleanupJob) Run(ctx context.Context) error {
+	const op = "controller.(sessionCleanupJob).Run"
+	j.numProcessed = 0
+	j.numWorkers = 0
+
+	// Load repos.
+	serversRepo, err := j.serversRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op, errors.WithMsg("error getting servers repo"))
+	}
+
+	sessionRepo, err := j.sessionRepoFn()
+	if err != nil {
+		return errors.Wrap(err, op, errors.WithMsg("error getting session repo"))
+	}
+
+	// Fetch all servers, ensure that time is not an issue
+	workers, err := serversRepo.ListServers(ctx, servers.ServerTypeWorker, servers.WithLiveness(-1))
+	if err != nil {
+		return errors.Wrap(err, op, errors.WithMsg("error listing workers"))
+	}
+
+	j.numWorkers = len(workers)
+	j.logger.Debug(
+		"starting job",
+		"op", op,
+		"num_processed", j.numProcessed,
+		"num_workers", j.numWorkers,
+	)
+	for _, worker := range workers {
+		j.logger.Debug(
+			"checking worker status",
+			"op", op,
+			"private_id", worker.PrivateId,
+			"num_processed", j.numProcessed,
+			"num_workers", j.numWorkers,
+		)
+
+		// Increment numProcessed here, as all other states are final.
+		j.numProcessed++
+
+		// Check update time against our grace period as a guard.
+		isPastGrace, updateTime, gracePeriod := j.isPastGrace(worker.UpdateTime.GetTimestamp().AsTime())
+		if !isPastGrace {
+			j.logger.Debug(
+				"worker status OK, reported within acceptable period",
+				"op", op,
+				"private_id", worker.PrivateId,
+				"update_time", updateTime,
+				"grace_period", gracePeriod,
+				"num_processed", j.numProcessed,
+				"num_workers", j.numWorkers,
+			)
+			continue
+		}
+
+		// If the guard failed, that means that our worker has not
+		// reported within the acceptable grace period. Find all active
+		// connections for the worker and close them.
+		numClosed, err := sessionRepo.CloseDeadConnectionsForWorker(ctx, worker.PrivateId, nil)
```

Comment:
I am wondering if it might be beneficial to move this call outside of the worker range, and then create a modified query that instead of calling:
```
-- Related to the worker that just reported to us
      server_id = $1
```
does a `server_id in` with a list of worker_ids that we collected in the range which were outside the grace period.

Reply:
@louisruch are you referring to having a single query versus more than one? If so, @jefferai and I discussed this and concluded that there would not be enough workers for this to matter.

Reply:
That is what I was suggesting.  Are we confident with the claim there won’t be enough workers to worry about?  This is something we can revisit in the future though, just felt like a relatively easy optimization.

Reply:
@jefferai thoughts? Do we have any data on what our upper limit on scale right now is for workers?

@louisruch I think that in terms of orders of magnitude, running a query for every _worker_ is probably much safer than running a query for, say, every _connection_. Probably within the acceptable range, but it would be good to know what our expect scale is for workers before saying 100% for sure.

Reply:
Given that the database holds the last updated time, I wonder if we even need to calculate each worker's grace individually. Think about a supermarket where they have a sign like, if the birth date is past today in 1974, that person can buy booze. I wonder if we really can and ought to use the DB for this. Basically, just tell the DB, if the last status time is before now minus 15 seconds, close all connections. I think this could generate the list and close things in a single CTE in a single transaction.

Reply:
@jefferai you're suggestion would seem to scale better and have the benefit of eliminating some race conditions as well.

Reply:
Alright, will move this to a single query. Incidentally, this was the original thought (when I was looking at table changes to make this happen), but it was scrapped when we decided that multiple queries was fine.

I'm sure we can do it without additional table changes though, so don't take that as me saying that we need to do that again.

Reply:
This is now done!